pkg/frontend/querymiddleware/labels_query_cache.go

```diff
@@ -0,0 +1,202 @@
+}
+
+type labelsQueryCache struct {
+	limits Limits
+}
+
+func (c *labelsQueryCache) getTTL(userID string) time.Duration {
+	return c.limits.ResultsCacheTTLForLabelsQuery(userID)
+}
+
+func (c *labelsQueryCache) parseRequest(req *http.Request) (*genericQueryRequest, error) {
+	if err := req.ParseForm(); err != nil {
+		return nil, err
+	}
+
+	var (
+		cacheKeyPrefix string
+		labelName      string
+	)
+
+	// Detect the request type
+	switch {
+	case strings.HasSuffix(req.URL.Path, labelNamesPathSuffix):
+		cacheKeyPrefix = labelNamesQueryCachePrefix
+	case labelValuesPathSuffix.MatchString(req.URL.Path):
+		cacheKeyPrefix = labelValuesQueryCachePrefix
+		labelName = labelValuesPathSuffix.FindStringSubmatch(req.URL.Path)[1]
+	default:
+		return nil, errors.New("unknown labels API endpoint")
+	}
+
+	// Both the label names and label values API endpoints support the same exact parameters (with the same defaults),
+	// so in this function there's no distinction between the two.
+	startTime, err := parseRequestTimeParam(req, "start", prometheusMinTime)
+	if err != nil {
+		return nil, err
+	}
+
+	endTime, err := parseRequestTimeParam(req, "end", prometheusMaxTime)
+	if err != nil {
+		return nil, err
+	}
+
+	matcherSets, err := parseRequestMatchersParam(req, "match[]")
+	if err != nil {
+		return nil, err
+	}
+
+	return &genericQueryRequest{
+		cacheKey:       generateLabelsQueryRequestCacheKey(startTime, endTime, labelName, matcherSets),
+		cacheKeyPrefix: cacheKeyPrefix,
+	}, nil
+}
+
+func generateLabelsQueryRequestCacheKey(startTime, endTime int64, labelName string, matcherSets [][]*labels.Matcher) string {
+	var (
+		twoHoursMillis = (2 * time.Hour).Milliseconds()
+		b              = strings.Builder{}
+	)
+
+	// Align start and end times to default block boundaries. The reason is that both TSDB (so the Mimir ingester)
+	// and Mimir store-gateway query the label names and values out of blocks overlapping within the start and end
+	// time. This means that for maximum granularity is the block.
+	if startTime != prometheusMinTime {
+		if reminder := startTime % twoHoursMillis; reminder != 0 {
+			startTime -= reminder
+		}
+	}
+	if endTime != prometheusMaxTime {
+		if reminder := endTime % twoHoursMillis; reminder != 0 {
+			endTime += twoHoursMillis - reminder
+		}
+	}
+
+	// Add start and end time.
+	b.WriteString(fmt.Sprintf("%d", startTime))
+	b.WriteRune(stringParamSeparator)
+	b.WriteString(fmt.Sprintf("%d", endTime))
+
+	// Add label name (if any).
+	if labelName != "" {
+		b.WriteRune(stringParamSeparator)
+		b.WriteString(labelName)
+	}
+
+	// Add matcher sets.
+	b.WriteRune(stringParamSeparator)
+	b.WriteString(fmt.Sprintf("%d", len(matcherSets)))
+
+	for _, set := range matcherSets {
+		b.WriteRune(stringParamSeparator)
+
+		for idx, matcher := range set {
+			if idx > 0 {
+				b.WriteRune(stringValueSeparator)
+			}
+			b.WriteString(matcher.String())
+		}
+	}
+
```

Comment:
There is `util.MultiMatchersStringer` which does the same as what you implemented in this function. Can we reuse it?

Reply:
I wasn't aware of that util. Thanks for pointing out. Adopted here: https://github.com/grafana/mimir/pull/5426/commits/f228944de4b79805a20d058016447981463c6d13